src/modules/jobs/jobs.service.ts

```diff
@@ -1,21 +1,63 @@
-import { Injectable, NotFoundException, UnauthorizedException, UseGuards } from '@nestjs/common';
+import { HttpStatus, Injectable, NotFoundException } from '@nestjs/common';
 import { InjectRepository } from '@nestjs/typeorm';
-import { User } from '../user/entities/user.entity';
 import { Repository } from 'typeorm';
-import { Job } from './entities/job.entity';
-import { JobDto } from './dto/job.dto';
+import { CustomHttpException } from '../../helpers/custom-http-filter';
 import { pick } from '../../helpers/pick';
-import { PaginationDto } from './dto/pagination.dto';
+import { User } from '../user/entities/user.entity';
+import { FindJobResponseDto } from './dto/find-job-response.dto';
+import { JobApplicationResponseDto } from './dto/job-application-response.dto';
+import { JobApplicationDto } from './dto/job-application.dto';
+import { JobDto } from './dto/job.dto';
+import { JobApplication } from './entities/job-application.entity';
+import { Job } from './entities/job.entity';
+import { isPassed } from './utils/helpers';
 
 @Injectable()
 export class JobsService {
   constructor(
     @InjectRepository(User)
     private readonly userRepository: Repository<User>,
     @InjectRepository(Job)
-    private readonly jobRepository: Repository<Job>
+    private readonly jobRepository: Repository<Job>,
+    @InjectRepository(JobApplication)
+    private readonly jobApplicationRepository: Repository<JobApplication>
   ) {}
 
+  async applyForJob(jobId: string, jobApplicationDto: JobApplicationDto): Promise<JobApplicationResponseDto> {
+    const job: FindJobResponseDto = await this.getJob(jobId);
+
+    const { is_deleted, deadline } = job.data;
+
+    if (is_deleted) {
+      throw new CustomHttpException('Job deleted', HttpStatus.NOT_FOUND);
+    }
+
+    if (isPassed(deadline)) {
+      throw new CustomHttpException('Job application deadline passed', HttpStatus.UNPROCESSABLE_ENTITY);
```

Comment:
better to use constants for error and success messages where possible

Reply:
I am lost. There was an agreement for Customhttpexception and filter to be used to avoid rewriting code. This has been implemented across the codebase.

Reply:
Done